[FLINK-15948][yarn] Enrich the warning log for YARN minimum allocation memory

wangyang0918 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Currently, we just check whether the JobManager/TaskManager is smaller than YARN minimum allocation memory. I think it should be enriched by checking whether it is a multiple of YARN minimum allocation memory.

Because YARN will always normalize the resource request by insuring that the requested memory is a multiple of minimum allocation. The minimum allocation could be configured in YARN ResourceManager via 'yarn.scheduler.minimum-allocation-mb'.

Refer [FLINK-15948](https://issues.apache.org/jira/browse/FLINK-15948) for more information.

## Brief change log

* Update the warning log


## Verifying this change

* Manually test and logs look like the following
```
2020-02-24 20:14:38,542 INFO  org.apache.flink.yarn.YarnClusterDescriptor                   - The configured JobManager memory is 1023 MB. YARN will allocate 1024 MB to make up an integer multiple of its minimum allocation memory (128 MB, configured via 'yarn.scheduler.minimum-allocation-mb'). The extra 1 MB may not be used by Flink.
2020-02-24 20:14:38,542 INFO  org.apache.flink.yarn.YarnClusterDescriptor                   - The configured TaskManager memory is 1568 MB. YARN will allocate 1664 MB to make up an integer multiple of its minimum allocation memory (128 MB, configured via 'yarn.scheduler.minimum-allocation-mb'). The extra 96 MB may not be used by Flink.
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

wangyang0918 commented:
@tillrohrmann Could you take a look? This is a very small change to enrich the logging.

xintongsong commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -521,11 +521,18 @@ private ClusterSpecification validateClusterResources(
+			LOG.warn("The {} is not a multiple of YARN minimum allocation memory({}), so some extra memory will be wasted. "
+				+ "Because YARN will always normalize the resource request by insuring that the requested memory is a multiple "
+				+ "of minimum allocation. The value of 'yarn.scheduler.minimum-allocation-mb' is '" + yarnMinAllocationMB + "'. "
+				+ "Please increase the specified memory size.", noteMem.toString(), yarnMinAllocationMB);
```
I would suggest to have a more concise log message. The current one mentions "multiple of YARN min allocation" and the configured value of yarnMinAllocationMB twice each, which is not necessary.

I would suggest the following:
`The configured {} memory is {} MB. YARN will allocate {} MB to make up an integer multiple of its minimum allocation memory ({} MB, configured via 'yarn.scheduler.minimum-allocation-mb'). The extra {} MB may not be used by Flink.`

xintongsong replied:
I would suggest to wrap the checking and the logging into one method, and call it twice for both JM and TM. The method could look like:
`logIfComponentMemNotIntegerMultipleOfYarnMinAllocation(String componentName, int componentMemoryMB, int yarnMinAllocationMB)`

wangyang0918 commented:
@xintongsong Thanks a lot for your review and suggestion, especially for the text. I have addressed your comments. Please take a look at your convenience.

@tillrohrmann Do you mind to take a look too and help to merge?

wangyang0918 commented:
Merged to master via [d6f039](https://github.com/apache/flink/commit/d6f0396b0e171be1942d12ac77afa73852ab2809). Thanks @tillrohrmann.

Jiayi-Liao commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -566,6 +562,22 @@ private ClusterSpecification validateClusterResources(
+			String componentName,
+			int componentMemoryMB,
+			int yarnMinAllocationMB) {
+		int normalizedMemMB = (componentMemoryMB + (yarnMinAllocationMB - 1)) / yarnMinAllocationMB * yarnMinAllocationMB;
```
What if `yarnMinAllocationMB` equals to zero, which is the default value set in https://github.com/apache/flink/blob/03b82f9aa81b828b06c61f7f30f92a331ec5120b/flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java#L511

wangyang0918 replied:
Thanks for your comments. I think `yarnMinAllocationMB` will always be greater than 0(the default is 1024m). Otherwise, the Yarn resourcemanager could not start successfully.

You could verify this via adding or removing the `yarn.scheduler.minimum-allocation-mb` in your `yarn-site.xml`.

Jiayi-Liao replied:
Actually I'm here because I met the problem in our environment. I'm not an expert on YARN but I think we should not rely on the yarn's limitations here. Besides, if our assumption is that `yarnMinAllocationMB` should always be greater than 0, why do we set the default value as 0, which is obviously a wrong value and certainly crashes the job.

wangyang0918 replied:
Yes, i think it is a wrong a default value. We should use `YarnConfiguration.DEFAULT_RM_SCHEDULER_MINIMUM_ALLOCATION_MB` instead.

So in your environment, the `yarn.scheduler.minimum-allocation-mb` is configured to 0 explicitly? Otherwise, i am curious how it could happen.

Jiayi-Liao replied:
@wangyang0918 Yes. It's explicitely set to 0. And I don't know why...

Jiayi-Liao replied:
@wangyang0918 BTW, I can fix this today if you don't mind.

wangyang0918 replied:
Even i still believe that it should not be configured to 0 from Yarn side. However, i think Flink could do more things to avoid this.

Feel free to open a ticket. We could make the following two minor changes.
* Make the default value to `YarnConfiguration.DEFAULT_RM_SCHEDULER_MINIMUM_ALLOCATION_MB`, not 0
* Add pre-check for `yarnMinAllocationMB`, it should be greater than 0.

wangyang0918 replied:
In your case, you still need to configure `yarn.scheduler.minimum-allocation-mb` to a positive number(i.e. greater than 0). It is a mandatory requirement of Yarn.

Jiayi-Liao replied:
@wangyang0918 Yes. The change is not for my case, just make the default value here more sensible.
